Sources/SwiftFormat/Utilities/FileIterator.swift

```diff
@@ -97,12 +97,12 @@ public struct FileIterator: Sequence, IteratorProtocol {
           output = next
         }
       }
-      if let out = output, visited.contains(out.absoluteURL.standardized.path) {
+      if let out = output, visited.contains(out.absoluteURL.standardized.standardizedFileURL.path) {
```

Comment:
Do we still need `.standardized` or is this included in `standardizedFileURL`?

Same below.

Reply:
Oh, interesting. Could you add a test case for this so we don’t carelessly remove `standardized` in the future?

Reply:
Sorry, I couldn't add a test case because it doesn't matter in swift-format.

I found that the code example above was bad. The reason that `URL(filePath: "..").standardized.path` was `"/home/kebo"` was that the given URL was relative.

[`URL.standardized`](https://github.com/swiftlang/swift-foundation/blob/3f9171072dc842c04c77ed2fa578d1b0cd123a04/Sources/FoundationEssentials/URL/URL.swift#L1763-L1781) follows the following steps:

1. Create `components` from `_parseInfo` (but `_parseInfo` does not contain the `baseURL` info)
1. Assign `components.percentEncodedPath.removingDotSegments` to `components.percentEncodedPath`
1. Return `components.url(relativeTo: baseURL)!`

In step 2, `".."` will be `""` because `URLComponents` doesn't have the `baseURL` info and `removingDotSegments` will never append dots. [This test case](https://github.com/swiftlang/swift-foundation/blob/3f9171072dc842c04c77ed2fa578d1b0cd123a04/Tests/FoundationEssentialsTests/StringTests.swift#L424) also indicates that. Because of that, `removingDotSegments` can't go up above the `baseURL`. So in step 3, it will return a `URL` just like `URL(filePath: "", relativeTo: baseURL)`, so its `.path` equals to `baseURL.path`. Therefore, `URL(filePath: "..").standardized.path` became `"/home/kebo"`.

However, if the given URL is absolute, `removingDotSegments` can go up above the base URL because all information including the base URL is included in `URLComponents`. In addition, `baseURL` will be `nil`. So `URL(filePath: "..").absoluteURL.standardized.path` will be `"/home"`, and it's the same as `URL(filePath: "..").standardizedFileURL.path`.

In swift-format's `FileIterator`, `standardized` always follows `absoluteURL`. So I can assure that what I said [before](https://github.com/swiftlang/swift-format/pull/837#discussion_r1803669289) does not occur in swift-format.

----

So I looked through `URL.standardizedFileURL` once again.

- [`URL.standardizedFileURL`](https://github.com/swiftlang/swift-foundation/blob/3f9171072dc842c04c77ed2fa578d1b0cd123a04/Sources/FoundationEssentials/URL/URL.swift#L1789-L1799)
- [`String.standardizingPath`](https://github.com/swiftlang/swift-corelibs-foundation/blob/cb6b2fa46992e04ae3f77fae30e2d86cebd2b440/Sources/Foundation/NSPathUtilities.swift#L318-L329)
- [`String.expandingTildeInPath`](https://github.com/swiftlang/swift-corelibs-foundation/blob/cb6b2fa46992e04ae3f77fae30e2d86cebd2b440/Sources/Foundation/NSPathUtilities.swift#L284-L303)
- [`String.resolvingSymlinksInPath`](https://github.com/swiftlang/swift-corelibs-foundation/blob/cb6b2fa46992e04ae3f77fae30e2d86cebd2b440/Sources/Foundation/NSPathUtilities.swift#L331-L364)

`URL.standardizedFileURL` seems to be `expandingTildeInPath` + `resolvingSymlinksInPath` + removing the `/var/automount` prefix + making an absolute URL, and `resolvingSymlinksInPath` seems to include `URL.standardized` in addition to resolve symlinks and removing the `/private` prefix.

So I think it is safe to replace `.absoluteURL.standardized` with `.standardizedFileURL` at least on macOS and other Unix-like OSes. How about doing this instead of adding a test case for `.standardized`?

I have already tested the following diff can pass the all tests on macOS and Linux.

```diff
diff --git a/Sources/SwiftFormat/Utilities/FileIterator.swift b/Sources/SwiftFormat/Utilities/FileIterator.swift
index 8189f18..9d1a1d2 100644
--- a/Sources/SwiftFormat/Utilities/FileIterator.swift
+++ b/Sources/SwiftFormat/Utilities/FileIterator.swift
@@ -97,12 +97,12 @@ public struct FileIterator: Sequence, IteratorProtocol {
           output = next
         }
       }
-      if let out = output, visited.contains(out.absoluteURL.standardized.standardizedFileURL.path) {
+      if let out = output, visited.contains(out.standardizedFileURL.path) {
         output = nil
       }
     }
     if let out = output {
-      visited.insert(out.absoluteURL.standardized.standardizedFileURL.path)
+      visited.insert(out.standardizedFileURL.path)
     }
     return output
   }
```

Sorry for a very long post.

Reply:
Thanks for the detailed explanation, very nice read. What you suggest sounds like the best solution, let’s do that.

Reply:
I pushed b9065b18d8435618635e0f619e62a97e7dee55fb.